parsil/src/assembler.rs

```diff
@@ -470,19 +501,114 @@ impl<'a, C: ContextProvider> Assembler<'a, C> {
     /// to extract putative bounds on the secondary index. Only `AND`
     /// combinators are traversed, as any other one would not statically
     /// guarantee the predominance of the bound.
-    fn find_secondary_index_boundaries(&mut self, expr: &Expr) -> Result<()> {
-        self.maybe_set_secondary_index_boundary(expr)?;
+    fn find_secondary_index_boundaries(&mut self, expr: &Expr, bounds: &mut Bounds) -> Result<()> {
+        self.maybe_set_secondary_index_boundary(expr, bounds)?;
+        match expr {
+            Expr::BinaryOp {
+                left,
+                op: BinaryOperator::And,
+                right,
+            } => {
+                self.find_secondary_index_boundaries(left, bounds)?;
+                self.find_secondary_index_boundaries(right, bounds)?;
+                Ok(())
+            }
+            Expr::Nested(e) => self.find_secondary_index_boundaries(e, bounds),
+            _ => Ok(()),
+        }
+    }
+
+    /// If `expr` defines a definite boundary on the primary index, update the
+    /// adequate member of the `bounds` tuple.
+    fn maybe_set_primary_index_bounds(
+        &mut self,
+        expr: &Expr,
+        bounds: &mut (bool, bool),
+    ) -> Result<()> {
+        if let Expr::BinaryOp { left, op, right } = expr {
+            if self.is_primary_index(left)?
+                // SID can only be computed from constants and placeholders
+                && self.is_expr_static(right)?
+                // SID can only be defined by up to one level of BasicOperation
+                && Self::depth(right) <= 1
+            {
+                match op {
+                    // $PI > x
+                    BinaryOperator::Gt | BinaryOperator::GtEq => {
```

Comment:
$MIN_BLOCK and $MAX_BLOCK are meant to be inclusive range bounds for the circuits, so I think we should only accept `GtEq` and `LtEq' operators here, otherwise we will have a different semantic between the query and what is proven by the circuit. Wdyt?

Reply:
Ah right, I thought it was the other way around.

Reply:
If we meet a `Gt`, it means that the circuits will prove an additional row, but that would not make the proof invalid, would it?

Reply:
It will not invalidate the proof, but the circuit is always proving `>=` rather than `>`. So, if a user specifies `> $MIN_BLOCK`, I feel it's more appropriate to reject the query to avoid that the user is thinking to get a result for `>` rather than `>=`